### PR TITLE
better patch release doc

### PR DESCRIPTION
**Pull request by zburke**
It was pointed out to me that we don't have good documentation about how to prepare a patch release.

### CODE REVIEW

**Comment by VictorSoroka1** [comment]
Could you please clarify at what point the `Changelog` information about the patch release stuff is going to be added to the master branch?

File: doc/release-procedure.md [hunk 1]
```diff
@@ -83,6 +83,33 @@ In the Jira project, create a new version with this number, so that issues can b
 
 Create a new entry at the top of the change-log for the forthcoming version, so there is somewhere to add entries. But do not include a date for the entry: instead, mark it as "IN PROGRESS", as in [the in-progress `stripes-core` change-log from before v0.5.0](https://github.com/folio-org/stripes-core/blob/e058702cb19b32f607f7fb40b15ddf00cd6b45ad/CHANGELOG.md).
 
+## Patch release procedures
+
+In short, making a patch release is not any different than making a regular release, except that you create a new branch from the last-release's commit and everything happens relative to that branch, rather than relative to `master`. For example, if you are making the first patch to the `v3.4.0`, to be published as `v3.4.1`, create a `b3.4` branch and do everything relative to that:
+```
+git checkout -b b3.4 <release-commit-hash-of-v3.4.0>
```
Comment (MikeTaylor) [inline doc/release-procedure.md:90]: Why not just
```
git checkout -b b3.4 v3.4.0
```
? Shouldn't that release have been tagged?

File: doc/release-procedure.md [hunk 2]
```diff
@@ -83,6 +83,33 @@ In the Jira project, create a new version with this number, so that issues can b
...
+git push -u origin HEAD
+```
+For each ticket you want to include in the patch release, create a branch (as you normally would) and then use [`git cherry-pick`](https://git-scm.com/docs/git-cherry-pick) to copy those commits onto the branch. (Alternatively, you can `rebase` the original branch onto this patch branch if there are many commits and they were not squashed. `cherry-pick` is really just a manual `rebase`; in fact this is what Git does under the hood.) For example, if you are adding UIU-666 to the release:
```
Comment (MikeTaylor) [inline doc/release-procedure.md:93]: I'm losing track a bit here of what is "this branch" and "that branch". We have `v3.4.0`, `b3.4` and `UIU-666-patch` in the mix at this stage, right? I think these needs a bit more spelling out.

File: doc/release-procedure.md [hunk 3]
```diff
@@ -83,6 +83,33 @@ In the Jira project, create a new version with this number, so that issues can b
...
+```
+git checkout -b UIU-666-patch
+git cherry-pick <hash-of-commit-to-add>
```
Comment (MikeTaylor) [inline doc/release-procedure.md:96]: So are you saying to cherry-pick _from_ the patch-release branch _to_ the Jira-number branch? That is the opposite of what I would have expected.

**Comment by doytch** [comment]
This process is more complex than the one I use. I can't see any direct benefit to the various separate branches unless you feel it makes things simpler for a given developer. Feel free to go with whatever you want, but here's a writeup of my process.

-- 

This process assumes the fixes have already been committed to Git. The fixes should be in terse Git commits to avoid introducing unnecessary changes.

First, create (if necessary) and check out the branch for the release you're patching.
```
(master): git checkout -b b3.4 v3.4.0 #if a release branch doesn't exist
(master): git checkout b3.4 #if the release branch exists
```
Next, cherry-pick the commits that contain the fixes you want.
```
(b3.4): git cherry-pick <commit-sha-1>
(b3.4): git cherry-pick <commit-sha-2>
```
At this point, we can treat the branch like any release. In the `package.json`, update the `version` to `3.4.1` and add the new version information to `CHANGELOG.md`. Stage the files, create a release commit, tag the commit, and push everything up to Github.
```
(b3.4): vi package.json
(b3.4): vi CHANGELOG.md
(b3.4): git add package.json CHANGELOG.md
(b3.4): git commit -m "Release 3.4.1"
(b3.4): git tag v3.4.1
(b3.4): git push origin b3.4
(b3.4): git push --tags
```
Finally, visit Jenkins for the repo and start a build from this `v3.4.1` tag to publish the release to NPM.

File: doc/release-procedure.md (hunk 1 above)
Comment (zburke) [reply doc/release-procedure.md:90]: You're right; that's easier.

File: doc/release-procedure.md (hunk 3 above)
Comment (zburke) [reply doc/release-procedure.md:96]: Nope, my intention here was to say, "use git cherry-pick to copy the original commits (on `master`) onto your new Jira-number-branch `UIU-666-patch`, then merge that Jira-number-branch to your release-branch, `b3.4`. I'll try to make that more clear.

File: doc/release-procedure.md (hunk 2 above)
Comment (zburke) [reply doc/release-procedure.md:93]: So noted.

**Comment by zburke** [comment]
@doytch, I think the process I outlined is more complex because I am seldom able to rely on this fact: "The fixes should be in terse Git commits to avoid introducing unnecessary changes." So often, I am asked to make a patch release and the PR I am asked to include has a half-dozen commits and wasn't squashed when it was merged to master. Thus the extra song-and-dance of a `UIU-666-patch` branch rather than just directly cherry-picking the commit onto `b3.4`. Otherwise, I think the processes are basically the same. 

While writing this up, I went back and forth between a terse version that assumed the release branch `b3.4` already existed (because it should) and the PRs you're copying over would be in good shape, and the longer OMG branches-upon-branches version that's here. Maybe we need both? e.g. a TLDR section like what you outlined, plus a step-by-step "here's everything that could go wrong" version like what I wrote?

**Comment by doytch** [comment]
@zburke , ahhh, makes sense. So you're using the intermediate branch to create a squashed commit essentially?

Separate from that, I've changed my mind. While I always keep the release branch in place, I think it probably makes sense to just not keep it. Trying to reuse it invites issues with merge commits polluting and complicating a rebase/merge after the release is made. Eg, if you make a fix branch, release the hotfix, and then want to merge everything to `master`. I've seen both Mike and Jason comment that you can obviously just create a new branch off of the release tag and I've just never updated my methodology bc of fear of unknowns I think :P

**Comment by zburke** [comment]
@doytch: 
>  So you're using the intermediate branch to create a squashed commit essentially?

Exactly. But your write up and formatting of the Platonic Ideal patch release was so clear and concise that I just stole it and, as you suggested, add a note about using extra branches if the commits are messy. Take another look? 

@VictorSoroka1: I also added a note about copying the patch-release-notes onto the master branch's `CHANGELOG.md` to help remind folks to do that.

**Comment by zburke** [comment]
Thanks for catching the out-of-sync ToC, @dcrossleyau!

**Pull request merged**